versioned/storage/cleanup/src/test/java/org/projectnessie/versioned/storage/cleanup/TestCutHistory.java

```diff
@@ -0,0 +1,310 @@
+    return IcebergTable.of(key, 1, 2, 3, 4);
+  }
+
+  private Hash commit(VersionStore store, String ref, Collection<String> keys) throws Exception {
+    ReferenceInfo<CommitMeta> main = store.getNamedRef(ref, GetNamedRefsParams.DEFAULT);
+    CommitResult result =
+        store.commit(
+            BranchName.of(main.getNamedRef().getName()),
+            Optional.of(main.getHash()),
+            CommitMeta.fromMessage("test: " + keys),
+            keys.stream()
+                .map(k -> Operation.Put.of(ContentKey.of(k), table(k)))
+                .collect(Collectors.toList()));
+
+    return result.getCommit().getHash();
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {1, 10, 100, 500})
+  void cutPreservesContents(int numExtraKeys) throws Exception {
+    soft.assertThat(repositoryLogic(persist).repositoryExists()).isTrue();
+
+    // Cut point: C1
+    //           /- G3 ---------\
+    //          /- E1 - E2 -\    \ /- F1
+    //  root - C0 --------- C1 - C2 - B1... Bnn
+    //                       \- D1 - D2
+    var main =
+        requireNonNull(repositoryLogic(persist).fetchRepositoryDescription()).defaultBranchName();
+    VersionStore store = new VersionStoreImpl(persist);
+    var root = commit(store, main, List.of("root"));
+
+    Set<String> c0Keys =
+        IntStream.rangeClosed(1, numExtraKeys).mapToObj(i -> "k" + i).collect(Collectors.toSet());
+    var c0 = commit(store, main, c0Keys); // force index stripes to still out
+
+    var branchE = store.create(BranchName.of("e"), Optional.of(c0)).getNamedRef().getName();
+    var e1 = commit(store, branchE, List.of("e1"));
+    var e2 = commit(store, branchE, List.of("e2"));
+
+    var branchG = store.create(BranchName.of("g"), Optional.of(c0)).getNamedRef().getName();
+    var g1 = commit(store, branchG, List.of("g1"));
+
+    var c1 =
+        store
+            .merge(
+                VersionStore.MergeOp.builder()
+                    .expectedHash(Optional.of(c0))
+                    .fromRef(BranchName.of(branchE))
+                    .fromHash(e2)
+                    .toBranch(BranchName.of(main))
+                    .build())
+            .getResultantTargetHash();
+
+    var branchD = store.create(BranchName.of("d"), Optional.ofNullable(c1)).getNamedRef().getName();
+    var d1 = commit(store, branchD, List.of("d1"));
+    var d2 = commit(store, branchD, List.of("d2"));
+
+    var c2 =
+        store
+            .merge(
+                VersionStore.MergeOp.builder()
+                    .expectedHash(Optional.of(c0))
+                    .fromRef(BranchName.of(branchG))
+                    .fromHash(g1)
+                    .toBranch(BranchName.of(main))
+                    .build())
+            .getResultantTargetHash();
+
+    var branchF = store.create(BranchName.of("f"), Optional.ofNullable(c2)).getNamedRef().getName();
+    var f1 = commit(store, branchF, List.of("f1"));
+
+    List<String> keysB = new ArrayList<>();
```

Comment:
Nit: If you go all in for `var`, this can be rewritten as
```java
var keysB = new ArrayList<String>();
```

But I do not mind to mix both ways.